refactor: Extract util/fs from util/system

sedited opened the pull request:
This pull request is part of the `libbitcoinkernel` project https://github.com/bitcoin/bitcoin/issues/24303 https://github.com/bitcoin/bitcoin/projects/18 and more specifically its "Step 2: Decouple most non-consensus code from libbitcoinkernel". This commit was originally authored by empact and is taken from its parent PR #25152.

#### Context

There is an ongoing effort to decouple the `ArgsManager` used for command line parsing user-provided arguments from the libbitcoinkernel library (https://github.com/bitcoin/bitcoin/pull/25290, https://github.com/bitcoin/bitcoin/pull/25487, https://github.com/bitcoin/bitcoin/pull/25527, https://github.com/bitcoin/bitcoin/pull/25862, https://github.com/bitcoin/bitcoin/pull/26177, and https://github.com/bitcoin/bitcoin/pull/27125). The `ArgsManager` is defined in `system.h`. A similar pull request extracting functionality from `system.h` has been merged in https://github.com/bitcoin/bitcoin/pull/27238.

#### Changes

Next to providing better code organization, this PR removes some reliance of the tree of libbitcoinkernel header includes on `system.h` (and thus the `ArgsManager` definition) by moving filesystem related functions out of the `system.*` files. 

There is already a pair of `fs.h` / `fs.cpp` in the top-level `src/` directory. They were not combined with the files introduced here, to keep the patch cleaner and more importantly because they are often included without the utility functions. The new files are therefore named `fs_helpers` and the existing `fs` files are moved into the util directory.

Further commits splitting more functionality out of `system.h` are still in #25152 and will be submitted in separate PRs once this PR has been processed.

Empact committed:
refactor: Extract util/fs_helpers from util/system

This is an extraction of filesystem related functions from util/system
into their own utility file.

The background of this commit is an ongoing effort to decouple the
libbitcoinkernel library from the ArgsManager defined in system.h.
Moving these functions out of system.h allows including them from a
separate source file without including the ArgsManager definitions from
system.h.

fanquake commented on src/util/fs.h:
```diff
@@ -0,0 +1,63 @@
+fs::path GetSpecialFolderPath(int nFolder, bool fCreate = true);
+#endif
+
+#endif // BITCOIN_UTIL_FS_H
```
nit: missing newline

TheCharlatan commented on src/flatfile.cpp:
```diff
@@ -8,12 +8,11 @@
-    m_dir(std::move(dir)),
-    m_prefix(prefix),
-    m_chunk_size(chunk_size)
+FlatFileSeq::FlatFileSeq(fs::path dir, const char* prefix, size_t chunk_size) : m_dir(std::move(dir)),
```
This is from clang format diff. Should I drop these kind of changes in the future? There are also some other not strictly related include order fixes.

fanquake replied:
I think include fixing is fine. However theres no need to include unrelated clang-format changes.

TheCharlatan replied:
I'm formatting each commit with `git diff -U0 HEAD~1.. | ./contrib/devtools/clang-format-diff.py -p1 -i -v`. I think I must have included this unrelated change somewhere. Will fix.

fanquake commented on src/test/util_tests.cpp:
```diff
@@ -2,7 +2,7 @@
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
-#include <util/system.h>
+#include <util/fs.h>
```
Pretty sure this can just be merged into the includes below.

TheCharlatan commented:
Thanks for the review!

Updated b67efe778d4e4ad359fb8dccea7f92e17d1b1f95 -> ea278094e9937bf96157161941d38a0c08a0aa4e ([splitSystemFs_0](https://github.com/TheCharlatan/bitcoin/commits/splitSystemFs_0) -> [splitSystemFs_1](https://github.com/TheCharlatan/bitcoin/commits/splitSystemFs_1), [compare](https://github.com/TheCharlatan/bitcoin/compare/splitSystemFs_0..splitSystemFs_1)) to address review comments.

ryanofsky commented:
Agree these functions should not be combined into existing `src/fs.h` since they are doing higher level things than accessing the filesystem. But instead of having two `fs.h` files in different locations, I would move `src/fs.h` to `src/util/fs.h` and move these functions to `src/util/fs_misc.h` or another place like that.

TheCharlatan commented:
Updated ea278094e9937bf96157161941d38a0c08a0aa4e -> c4f2b6d619f6c0faee5aa3118ade1040105fdd8b ([splitSystemFs_1](https://github.com/TheCharlatan/bitcoin/commits/splitSystemFs_1) -> [splitSystemFs_2](https://github.com/TheCharlatan/bitcoin/commits/splitSystemFs_2), [compare](https://github.com/TheCharlatan/bitcoin/compare/splitSystemFs_1..splitSystemFs_2)) to rename `util/fs.*` to `util/fs_helpers.*`

Added commit dac4d1dc3d75d510efe6c84c2cb6d7d7dc96e2c6 ([splitSystemFs_3](https://github.com/TheCharlatan/bitcoin/commits/splitSystemFs_3)) as a scripted diff moving `src/fs.*` to `src/util/fs.*` as requested by @ryanofsky in https://github.com/bitcoin/bitcoin/pull/27254#issuecomment-1468037295 .

sedited committed:
refactor: Move fs.* to util/fs.*

The fs.* files are already part of the libbitcoin_util library. With the
introduction of the fs_helpers.* it makes sense to move fs.* into the
util/ directory as well.

TheCharlatan commented:
Updated dac4d1dc3d75d510efe6c84c2cb6d7d7dc96e2c6 -> b22cf8d563a469e29c9942b4fd9f93351d8b9766 ([splitSystemFs_3](https://github.com/TheCharlatan/bitcoin/commits/splitSystemFs_3) -> [splitSystemFs_4](https://github.com/TheCharlatan/bitcoin/commits/splitSystemFs_4), [compare](https://github.com/TheCharlatan/bitcoin/compare/splitSystemFs_3..splitSystemFs_4)) to fix a broken rename and include an additional header in `compat/assumptions.h`. Up until this point `compat/assumptions.h` relied on being included after `compat/compat.h`. `clang-format-diff` revealed this by changing the order of the includes. I'm not sure if this inclusion order has to be preserved.

sedited committed:
Add missing cstddef include in assumptions.h

The inclusion of this header should not depend on the inclusion of other
headers that include cstddef themselves.

sedited committed:
Add missing fs.h includes

The inclusion of this header should not depend on the inclusion of other
headers that include fs.h themselves.

TheCharlatan commented:
Rebased b22cf8d563a469e29c9942b4fd9f93351d8b9766 -> cc662d09386e50eb92a7ccbaa1edff62fd8cdd44 ([splitSystemFs_4](https://github.com/TheCharlatan/bitcoin/commits/splitSystemFs_4) -> [splitSystemFs_5](https://github.com/TheCharlatan/bitcoin/commits/splitSystemFs_5), [compare](https://github.com/TheCharlatan/bitcoin/compare/splitSystemFs_4..splitSystemFs_5)) to fix conflicts and update the commit message to exclude the scripted diff.

hebasto commented:
Concept ACK.

hebasto commented on ci/test/06_script_b.sh:
```diff
@@ -70,6 +70,7 @@ if [ "${RUN_TIDY}" = "true" ]; then
           " src/util/check.cpp"\
           " src/util/error.cpp"\
           " src/util/exception.cpp"\
+          " src/util/fs.cpp"\
```
3665c1bdf7b9d47ccab39916c13a8cd9dfbf0eea

There is no `src/util/fs.cpp` file at this commit. It is renamed in the last commit only.

hebasto reviewed (approved):
ACK 3ceb8dde48fc12f4f16372f661a4cccf7104e194

The commit "_Add missing fs.h include..._" could address another IWYU [suggestion](https://api.cirrus-ci.com/v1/task/6080090677706752/logs/ci.log):
```diff
--- a/src/node/chainstate.cpp
+++ b/src/node/chainstate.cpp
@@ -16,6 +16,7 @@
 #include <tinyformat.h>
 #include <txdb.h>
 #include <uint256.h>
+#include "util/fs.h"
 #include <util/time.h>
 #include <util/translation.h>
 #include <validation.h>
```

ryanofsky commented on src/util/fs_helpers.cpp:
```diff
@@ -0,0 +1,284 @@
+{
+    WCHAR pszPath[MAX_PATH] = L"";
+
+    if (SHGetSpecialFolderPathW(nullptr, pszPath, nFolder, fCreate)) {
```
In commit "refactor: Extract util/fs_helpers from util/system" (59c02e68f75f7b727162d2ea36bf3ed99fabf035)

Anybody have advice on easiest way to review changes like this when blocks of code are moving but also changing slightly at the same time?

I find it pretty easy to review blocks of moved code in git with the `--color-moved` option, but once code is moved and reformatted at the same time (the `{` at the end of this line was previously on its own line) it becomes very tedious to find the matching blocks of old and new code, and then compare them to make sure the differences are not significant.

If there isn't some more convenient way to review changes like this, I would definitely appreciate it if the code could just be moved and not reformatted.

TheCharlatan replied:
> If there isn't some more convenient way to review changes like this, I would definitely appreciate it if the code could just be moved and not reformatted.

ACK, I will stop using `clang-format-diff` on all moved code. Thank you for the review!

ryanofsky reviewed (approved):
Code review ACK 3ceb8dde48fc12f4f16372f661a4cccf7104e194

hebasto commented on src/qt/intro.cpp:
```diff
@@ -7,15 +7,16 @@
 #endif
 
 #include <chainparams.h>
-#include <fs.h>
 #include <qt/intro.h>
 #include <qt/forms/ui_intro.h>
+#include <util/fs.h>
 
 #include <qt/guiconstants.h>
 #include <qt/guiutil.h>
 #include <qt/optionsmodel.h>
 
 #include <interfaces/node.h>
+#include <util/fs_helpers.h>
 #include <util/system.h>
 #include <validation.h>
```
nit: 

```diff
--- a/src/qt/intro.cpp
+++ b/src/qt/intro.cpp
@@ -6,16 +6,16 @@
 #include <config/bitcoin-config.h>
 #endif
 
-#include <chainparams.h>
 #include <qt/intro.h>
 #include <qt/forms/ui_intro.h>
-#include <util/fs.h>
 
 #include <qt/guiconstants.h>
 #include <qt/guiutil.h>
 #include <qt/optionsmodel.h>
 
+#include <chainparams.h>
 #include <interfaces/node.h>
+#include <util/fs.h>
 #include <util/fs_helpers.h>
 #include <util/system.h>
 #include <validation.h>
```

hebasto reviewed (approved):
ACK 00e9b97f37e0bdf4c647236838c10b68b7ad5be3
